python/SWESimulators/BaseOceanStateEnsemble.py

```diff
@@ -0,0 +1,481 @@
+            print "Max eta: ", np.max(eta)
+            print "Max hu:  ", np.max(hu)
+            print "Max hv:  ", np.max(hv)
+            simNo = simNo + 1
+    
+
+                    
+            
+    def getEnsembleMean(self):
+        return None
+    def getDomainSizeX(self):
+        return self.nx*self.dx
+    def getDomainSizeY(self):
+        return self.ny*self.dy
+    def getObservationVariance(self):
+        return self.observation_variance
+    def getNumParticles(self):
+        return self.numParticles
+    
+    def findLargestPossibleTimeStep(self):
+        """
+        Un-optimized utility function to check largest allowed time-step for 
+        the shallow water model.
+        dt < 0.25*min[ dx/max(|u +- sqrt(g(H+eta))|), dy/max(|v +- sqrt(g(H+eta))|)]
+        """
+        max_u = 0
+        max_v = 0
+        for oceanState in self.particles:
+            eta_tmp, hu, hv = oceanState.download()
+            Hi = oceanState.downloadBathymetry()[0]
+            w = 0.25*(Hi[1:,1:] + Hi[1:,:-1] + Hi[:-1,1:] + Hi[:-1,:-1]) + eta_tmp
+            hu /= w
+            hv /= w
+            Hi = None
+            w = np.sqrt(self.g*w)
+            
+            # using eta_tmp buffer for {u|v} +- sqrt(gw)
+            max_u = max(max_u, np.max(np.abs(hu + w)))
+            max_u = max(max_u, np.max(np.abs(hu - w)))
+            max_v = max(max_v, np.max(np.abs(hv + w)))
+            max_v = max(max_v, np.max(np.abs(hv - w)))
+            
+        return 0.25*min(self.dx/max_u, self.dy/max_v)
+            
+            
+    
+    
+    def downloadEnsembleStatisticalFields(self):
+        eta_true, hu_true, hv_true = self.downloadTrueOceanState()
+        
+        eta_mean = np.zeros_like(eta_true)
+        hu_mean = np.zeros_like(hu_true)
+        hv_mean = np.zeros_like(hv_true)
+        eta_mrse = np.zeros_like(eta_true)
+        hu_mrse = np.zeros_like(hu_true)
+        hv_mrse = np.zeros_like(hv_true)
+        
+        for p in range(self.getNumParticles()):
+            tmp_eta, tmp_hu, tmp_hv = self.downloadParticleOceanState(p)
+            eta_mean += tmp_eta
+            hu_mean += tmp_hu
+            hv_mean += tmp_hv
+            eta_mrse += (eta_true - tmp_eta)**2
+            hu_mrse += (hu_true - tmp_hu)**2
+            hv_mrse += (hv_true - tmp_hv)**2
+            
+        eta_mean = eta_mean/self.getNumParticles()
+        hu_mean = hu_mean/self.getNumParticles()
+        hv_mean = hv_mean/self.getNumParticles()
+        eta_mrse = np.sqrt(eta_mrse/self.getNumParticles())
+        hu_mrse = np.sqrt(hu_mrse/self.getNumParticles())
+        hv_mrse = np.sqrt(hv_mrse/self.getNumParticles())
+        
+        return eta_mean, hu_mean, hv_mean, eta_mrse, hu_mrse, hv_mrse
+    
+    def downloadParticleOceanState(self, particleNo):
+        assert(particleNo < self.getNumParticles()+1), "particle out of range"
+        eta, hu, hv = self.particles[particleNo].download()
+        eta = eta[2:-2, 2:-2]
```

Comment:
Not that easy, since sim.ghostCells does not consider the "sponge zone" for open boundary conditions... Perhaps sim.download() should have an optional input parameter to specify whether ghost cells should be included or not?